File: Tests/KeyKeyUserDBKitTests/GramTests.swift
// (c) 2025 and onwards The vChewing Project (LGPL v3.0 License or later).
// ====================
// This code is released under the SPDX-License-Identifier: `LGPL-3.0-or-later`.

import Foundation
import Testing

@testable import KeyKeyUserDBKit

@Suite("Gram Tests")
struct GramTests {
  // MARK: - Creation Tests

  @Test("Create Unigram should set correct properties")
  func createUnigram_ShouldSetCorrectProperties() {
    let keyArray = ["ㄋㄧˇ", "ㄏㄠˇ"]
    let current = "你好"
    let probability = 0.5

    let gram = KeyKeyUserDBKit.Gram(
      keyArray: keyArray,
      current: current,
      probability: probability
    )

    #expect(gram.keyArray == keyArray)
    #expect(gram.current == current)
    #expect(gram.previous == nil)
    #expect(gram.probability == probability)
    #expect(gram.isCandidateOverride == false)
    #expect(gram.isUnigram == true)
  }

  @Test("Create Bigram should set correct properties")
  func createBigram_ShouldSetCorrectProperties() {
    let keyArray = ["ㄏㄠˇ"]
    let current = "好"
    let previous = "你"

    let gram = KeyKeyUserDBKit.Gram(
      keyArray: keyArray,
      current: current,
      previous: previous
    )

    #expect(gram.keyArray == keyArray)
    #expect(gram.current == current)
    #expect(gram.previous == previous)
    #expect(gram.probability == 0)
    #expect(gram.isCandidateOverride == false)
    #expect(gram.isUnigram == false)
  }

  @Test("Create Bigram with empty previous should set previous to nil")
  func createBigram_WithEmptyPrevious_ShouldSetPreviousToNil() {
    let gram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄏㄠˇ"],
      current: "好",
      previous: ""
    )
    #expect(gram.previous == nil)
    #expect(gram.isUnigram == true)
  }

  @Test("Create CandidateOverride should set correct properties")
  func createCandidateOverride_ShouldSetCorrectProperties() {
    let keyArray = ["ㄋㄧˇ"]
    let current = "妳"
    let probability = 114.514

    let gram = KeyKeyUserDBKit.Gram(
      keyArray: keyArray,
      current: current,
      probability: probability,
      isCandidateOverride: true
    )

    #expect(gram.keyArray == keyArray)
    #expect(gram.current == current)
    #expect(gram.previous == nil)
    #expect(gram.probability == probability)
    #expect(gram.isCandidateOverride == true)
  }

  // MARK: - Property Tests

  @Test("isReadingMismatched should return true when lengths differ")
  func isReadingMismatched_WhenLengthsDiffer_ShouldReturnTrue() {
    let gram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ", "ㄏㄠˇ"],
      current: "你好嗎" // 2 keys, 3 chars
    )
    #expect(gram.isReadingMismatched == true)
  }

  @Test("isReadingMismatched should return false when lengths match")
  func isReadingMismatched_WhenLengthsMatch_ShouldReturnFalse() {
    let gram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ", "ㄏㄠˇ"],
      current: "你好" // 2 keys, 2 chars
    )
    #expect(gram.isReadingMismatched == false)
  }

  @Test("segLength should return keyArray length")
  func segLength_ShouldReturnKeyArrayLength() {
    let gram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ", "ㄏㄠˇ", "ㄇㄚ"],
      current: "你好嗎"
    )
    #expect(gram.segLength == 3)
  }

  @Test("asTuple should return correct tuple")
  func asTuple_ShouldReturnCorrectTuple() {
    let gram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄏㄠˇ"],
      current: "好",
      previous: "你"
    )
    let tuple = gram.asTuple

    #expect(tuple.keyArray == ["ㄏㄠˇ"])
    #expect(tuple.value == "好")
    #expect(tuple.probability == 0)
    #expect(tuple.previous == "你")
  }

  // MARK: - Description Tests

  @Test("descriptionSansReading for Unigram should format correctly")
  func descriptionSansReading_Unigram_ShouldFormatCorrectly() {
    let gram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ"],
      current: "你",
      probability: 0.5
    )
    #expect(gram.descriptionSansReading == "P(你)=0.5")
  }

  @Test("descriptionSansReading for Bigram should format correctly")
  func descriptionSansReading_Bigram_ShouldFormatCorrectly() {
    let gram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄏㄠˇ"],
      current: "好",
      previous: "你"
    )
    #expect(gram.descriptionSansReading == "P(好|你)=0.0")
  }

  @Test("descriptionSansReading for CandidateOverride should format correctly")
  func descriptionSansReading_CandidateOverride_ShouldFormatCorrectly() {
    let gram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ"],
      current: "妳",
      probability: 114.514,
      isCandidateOverride: true
    )
    #expect(gram.descriptionSansReading == "P(妳)")
  }

  @Test("describe should include header and body")
  func describe_ShouldIncludeHeaderAndBody() {
    let gram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ"],
      current: "你",
      probability: 0.5
    )
    let description = gram.describe(keySeparator: "-")

    #expect(description.contains("[Unigram]"))
    #expect(description.contains("ㄋㄧˇ"))
    #expect(description.contains("P(你)=0.5"))
  }

  @Test("description should call describe")
  func description_ShouldCallDescribe() {
    let gram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ"],
      current: "你",
      probability: 0.5
    )
    #expect(gram.description == gram.describe(keySeparator: "-"))
  }

  // MARK: - Equality Tests

  @Test("Equality with same values should be equal")
  func equality_SameValues_ShouldBeEqual() {
    let gram1 = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ"],
      current: "你",
      probability: 0.5
    )
    let gram2 = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ"],
      current: "你",
      probability: 0.5
    )

    #expect(gram1 == gram2)
    #expect(gram1.hashValue == gram2.hashValue)
  }

  @Test("Equality with different values should not be equal")
  func equality_DifferentValues_ShouldNotBeEqual() {
    let gram1 = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ"],
      current: "你",
      probability: 0.5
    )
    let gram2 = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ"],
      current: "妳",
      probability: 0.5
    )

    #expect(gram1 != gram2)
  }

  // MARK: - Codable Tests

  @Test("Codable encode/decode should preserve values")
  func codable_EncodeDecode_ShouldPreserveValues() throws {
    let originalGram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ", "ㄏㄠˇ"],
      current: "你好",
      previous: "大家",
      probability: 0.75,
      isCandidateOverride: false
    )

    let encoder = JSONEncoder()
    let data = try encoder.encode(originalGram)

    let decoder = JSONDecoder()
    let decodedGram = try decoder.decode(KeyKeyUserDBKit.Gram.self, from: data)

    #expect(originalGram == decodedGram)
    #expect(originalGram.isCandidateOverride == decodedGram.isCandidateOverride)
  }

  @Test("Codable for CandidateOverride should preserve flag")
  func codable_CandidateOverride_ShouldPreserveFlag() throws {
    let originalGram = KeyKeyUserDBKit.Gram(
      keyArray: ["ㄋㄧˇ"],
      current: "妳",
      probability: 114.514,
      isCandidateOverride: true
    )

    let encoder = JSONEncoder()
    let data = try encoder.encode(originalGram)

    let decoder = JSONDecoder()
    let decodedGram = try decoder.decode(KeyKeyUserDBKit.Gram.self, from: data)

    #expect(decodedGram.isCandidateOverride == true)
  }

  // MARK: - Raw Tuple Constructor Tests

  @Test("Raw tuple constructor should set correct properties")
  func rawTupleConstructor_ShouldSetCorrectProperties() {
    let rawTuple: KeyKeyUserDBKit.Gram.GramRAW = (
      keyArray: ["ㄋㄧˇ"],
      value: "你",
      probability: 0.5,
      previous: "我"
    )

    let gram = KeyKeyUserDBKit.Gram(rawTuple)

    #expect(gram.keyArray == ["ㄋㄧˇ"])
    #expect(gram.current == "你")
    #expect(gram.previous == "我")
    #expect(gram.probability == 0.5)
    #expect(gram.isCandidateOverride == false)
  }

  @Test("Raw tuple constructor with CandidateOverride should set flag")
  func rawTupleConstructor_WithCandidateOverride_ShouldSetFlag() {
    let rawTuple: KeyKeyUserDBKit.Gram.GramRAW = (
      keyArray: ["ㄋㄧˇ"],
      value: "妳",
      probability: 114.514,
      previous: nil
    )

    let gram = KeyKeyUserDBKit.Gram(rawTuple, isCandidateOverride: true)

    #expect(gram.isCandidateOverride == true)
  }
}

File: WinNT/KeyKeyUserDBKit.Tests/GramTests.cs
// (c) 2025 and onwards The vChewing Project (LGPL v3.0 License or later).
// ====================
// This code is released under the SPDX-License-Identifier: `LGPL-3.0-or-later`.

using Xunit;

namespace KeyKeyUserDBKit.Tests;

public class GramTests {
  [Fact]
  public void CreateUnigram_ShouldSetCorrectProperties() {
    var keyArray = new[] { "ㄋㄧˇ", "ㄏㄠˇ" };
    var current = "你好";
    var probability = 0.5;

    var gram = Gram.CreateUnigram(keyArray, current, probability);

    Assert.Equal(keyArray, gram.KeyArray);
    Assert.Equal(current, gram.Current);
    Assert.Null(gram.Previous);
    Assert.Equal(probability, gram.Probability);
    Assert.False(gram.IsCandidateOverride);
    Assert.True(gram.IsUnigram);
  }

  [Fact]
  public void CreateBigram_ShouldSetCorrectProperties() {
    var keyArray = new[] { "ㄏㄠˇ" };
    var current = "好";
    var previous = "你";

    var gram = Gram.CreateBigram(keyArray, current, previous);

    Assert.Equal(keyArray, gram.KeyArray);
    Assert.Equal(current, gram.Current);
    Assert.Equal(previous, gram.Previous);
    Assert.Equal(0, gram.Probability);
    Assert.False(gram.IsCandidateOverride);
    Assert.False(gram.IsUnigram);
  }

  [Fact]
  public void CreateBigram_WithEmptyPrevious_ShouldSetPreviousToNull() {
    var gram = Gram.CreateBigram(["ㄏㄠˇ"], "好", "");
    Assert.Null(gram.Previous);
    Assert.True(gram.IsUnigram);
  }

  [Fact]
  public void CreateCandidateOverride_ShouldSetCorrectProperties() {
    var keyArray = new[] { "ㄋㄧˇ" };
    var current = "妳";
    var probability = 114.514;

    var gram = Gram.CreateCandidateOverride(keyArray, current, probability);

    Assert.Equal(keyArray, gram.KeyArray);
    Assert.Equal(current, gram.Current);
    Assert.Null(gram.Previous);
    Assert.Equal(probability, gram.Probability);
    Assert.True(gram.IsCandidateOverride);
  }

  [Fact]
  public void IsReadingMismatched_WhenLengthsDiffer_ShouldReturnTrue() {
    var gram = Gram.CreateUnigram(["ㄋㄧˇ", "ㄏㄠˇ"], "你好嗎"); // 2 keys, 3 chars
    Assert.True(gram.IsReadingMismatched);
  }

  [Fact]
  public void IsReadingMismatched_WhenLengthsMatch_ShouldReturnFalse() {
    var gram = Gram.CreateUnigram(["ㄋㄧˇ", "ㄏㄠˇ"], "你好"); // 2 keys, 2 chars
    Assert.False(gram.IsReadingMismatched);
  }

  [Fact]
  public void SegLength_ShouldReturnKeyArrayLength() {
    var gram = Gram.CreateUnigram(["ㄋㄧˇ", "ㄏㄠˇ", "ㄇㄚ"], "你好嗎");
    Assert.Equal(3, gram.SegLength);
  }

  [Fact]
  public void AsTuple_ShouldReturnCorrectTuple() {
    var gram = Gram.CreateBigram(["ㄏㄠˇ"], "好", "你");
    var tuple = gram.AsTuple;

    Assert.Equal(["ㄏㄠˇ"], tuple.KeyArray);
    Assert.Equal("好", tuple.Value);
    Assert.Equal(0, tuple.Probability);
    Assert.Equal("你", tuple.Previous);
  }

  [Fact]
  public void DescriptionSansReading_Unigram_ShouldFormatCorrectly() {
    var gram = Gram.CreateUnigram(["ㄋㄧˇ"], "你", 0.5);
    Assert.Equal("P(你)=0.5", gram.DescriptionSansReading);
  }

  [Fact]
  public void DescriptionSansReading_Bigram_ShouldFormatCorrectly() {
    var gram = Gram.CreateBigram(["ㄏㄠˇ"], "好", "你");
    Assert.Equal("P(好|你)=0", gram.DescriptionSansReading);
  }

  [Fact]
  public void DescriptionSansReading_CandidateOverride_ShouldFormatCorrectly() {
    var gram = Gram.CreateCandidateOverride(["ㄋㄧˇ"], "妳", 114.514);
    Assert.Equal("P(妳)", gram.DescriptionSansReading);
  }

  [Fact]
  public void Describe_ShouldIncludeHeaderAndBody() {
    var gram = Gram.CreateUnigram(["ㄋㄧˇ"], "你", 0.5);
    var description = gram.Describe();

    Assert.Contains("[Unigram]", description);
    Assert.Contains("ㄋㄧˇ", description);
    Assert.Contains("P(你)=0.5", description);
  }

  [Fact]
  public void ToString_ShouldCallDescribe() {
    var gram = Gram.CreateUnigram(["ㄋㄧˇ"], "你", 0.5);
    Assert.Equal(gram.Describe(), gram.ToString());
  }

  [Fact]
  public void Equality_SameValues_ShouldBeEqual() {
    var gram1 = Gram.CreateUnigram(["ㄋㄧˇ"], "你", 0.5);
    var gram2 = Gram.CreateUnigram(["ㄋㄧˇ"], "你", 0.5);

    Assert.True(gram1.Equals(gram2));
    Assert.Equal(gram1.GetHashCode(), gram2.GetHashCode());
  }

  [Fact]
  public void Equality_DifferentValues_ShouldNotBeEqual() {
    var gram1 = Gram.CreateUnigram(["ㄋㄧˇ"], "你", 0.5);
    var gram2 = Gram.CreateUnigram(["ㄋㄧˇ"], "妳", 0.5);

    Assert.False(gram1.Equals(gram2));
  }

  [Fact]
  public void Equality_WithNull_ShouldNotBeEqual() {
    var gram = Gram.CreateUnigram(["ㄋㄧˇ"], "你", 0.5);
    Assert.False(gram.Equals(null));
  }
}
